heat/core/dndarray.py

```diff
@@ -661,30 +661,34 @@ def balance_(self):
         send_slice = [slice(None), ] * self.numdims
         keep_slice = [slice(None), ] * self.numdims
         # need to send from the last one with data
-        for spr in range(last_pr_w_data, first_pr_w_data - 1, -1):
-            if self.comm.rank == spr:
-                for pr in range(self.comm.size - 1, spr, -1):
-                    send_amt = abs((chunk_map[pr, self.split] - lshape_map[pr, self.split]).item())
-                    send_amt = send_amt if send_amt < self.lshape[self.split] else self.lshape[self.split]
-                    if send_amt:
+        # start from x then push the data to the next one. then do the same at x+1 until the last process
+        balanced_process = [False for _ in range(self.comm.size)]
+        for pr in range(self.comm.size):
+            balanced_process[pr] = True if chunk_map[pr, self.split] == lshape_map[pr, self.split] else False
+            if pr > 0:
+                if any(i is False for i in balanced_process[:pr]):
+                    balanced_process[pr] = False
+
+        for pr, b in enumerate(balanced_process[:-1]):
+            if not b:  # if the process is not balanced
+                send_amt = abs((chunk_map[pr, self.split] - lshape_map[pr, self.split]).item())
+                send_amt = send_amt if send_amt < lshape_map[pr, self.split] else lshape_map[pr, self.split]
+                if send_amt:
+                    if self.comm.rank == pr:  # send data to the next process
                         send_slice[self.split] = slice(self.lshape[self.split] - send_amt, self.lshape[self.split])
                         keep_slice[self.split] = slice(0, self.lshape[self.split] - send_amt)
 
-                        self.comm.Isend(self.__array[send_slice].clone(), dest=pr, tag=pr + self.comm.size + spr)
+                        self.comm.Send(self.__array[send_slice].clone(), dest=pr + 1, tag=pr + self.comm.size + pr + 1)
```

Comment:
this isnt always the case that it _needs_ to be sent to the next rank but it is a save way to do it. it will cause a higher shuffle but this is safe (unless the data to be split between processes n+1 -> size -1 is much larger than is available on processes n+1)

Reply:
Right, but wouldn't we want to do this in parallel using an Alltoallv instead of an entirely sequential push-through?

Reply:
my goal was to avoid using Alltoallv to avoid the cost that it carries. I have the impression that in most cases i feel that it would be much heavier than is frequently required. if the sliced array is mostly balanced then an alltoall would be a superfluous call, however if thearray is very imbalanced than an alltoall makes more sense. what do you think? do you think that the balance function should just always call alltoall or is it better to have the logic hardcoded?

Reply:
Fair point. We should keep it then as is and potentially recode it using alltoallv later if necessary